src/zcash/address/transparent.h

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2021 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_ZCASH_ADDRESS_TRANSPARENT_H
+#define ZCASH_ZCASH_ADDRESS_TRANSPARENT_H
+
+#include "zip32.h"
+
+namespace libzcash {
+namespace transparent {
+
+class AccountPubKey {
+private:
+    CChainablePubKey pubkey;
+public:
+    AccountPubKey(CChainablePubKey pubkey): pubkey(pubkey) {};
+
+    const CChainablePubKey& GetPubKey() const {
+        return pubkey;
+    }
+
+    std::optional<CPubKey> DeriveExternal(uint32_t addrIndex) const;
+
+    std::optional<CPubKey> DeriveInternal(uint32_t addrIndex) const;
+
+    /**
+     * Return the transparent change address for this key a the given diversifier
+     * index, if that index is valid for a transparent address and if key
+     * derivation is successful.
+     */
+    std::optional<CKeyID> GetChangeAddress(const diversifier_index_t& j) const;
+
+    /**
+     * Search the space of valid child indices starting at the given index, and
+     * return the change address corresponding to the first index at which a
+     * valid address is derived, along with that index. This will always return
+     * a valid address unless the space of valid transparent child indices is
+     * exhausted or the provided diversifier index exceeds the maximum allowed
+     * non-hardened transparent child index.
+     */
+    std::optional<std::pair<CKeyID, diversifier_index_t>> FindChangeAddress(diversifier_index_t j) const;
+
+    /**
+     * Return the internal and external OVKs for shielding from transparent
+     * addresses derived from this key.
+     */
+    std::pair<uint256, uint256> GetOVKsForShielding() const;
+
+    friend bool operator==(const AccountPubKey& a, const AccountPubKey& b)
+    {
+        return a.pubkey == b.pubkey;
+    }
+};
+
+class AccountKey {
+private:
+    CExtKey accountKey;
+    CExtKey external;
+    CExtKey internal;
+
+    AccountKey(CExtKey accountKeyIn, CExtKey externalIn, CExtKey internalIn):
+        accountKey(accountKeyIn), external(externalIn), internal(internalIn) {}
+public:
+    static std::optional<AccountKey> ForAccount(
+            const HDSeed& mnemonic,
+            uint32_t bip44CoinType,
+            libzcash::AccountId accountId);
+
+    static HDKeyPath KeyPath(uint32_t bip44CoinType, libzcash::AccountId accountId) {
+        return
+            "m/44'/" +
+            std::to_string(bip44CoinType) + "'/" +
+            std::to_string(accountId) + "'";
+    }
+
+    static HDKeyPath KeyPath(uint32_t bip44CoinType, libzcash::AccountId accountId, bool external, uint32_t childIndex) {
```

Comment:
The meaning of `true` for the boolean flag is "external" here, and "internal" for `CWallet::GenerateNewKey`. Since both are new APIs, make them consistent (which avoids the need to invert the sense when calling `KeyPath` from `CWallet::GenerateNewKey`).